### ISSUE TITLE
[Bug] GLM-5.3-Flash-AWQ-W4A16 on sm89 (8xL20): deterministic prompt-token corruption + doubled tokens — evidence points to expert calibration (calibrate_all_experts=false)

### ISSUE THREAD

**Issue by xfdkuyi** [wtdcode/vllm-backport#102]
### Environment

- GPU: 8× NVIDIA L20 48GB (Ada, **sm89**), no NVLink
- Image: `lazymio/vllm-backport:latest-sm89` — engine reports v0.13.1, digest `0fc33cd7d2be` (**identical to the pinned `v0.13.1-sm89` tag**)
- Model: `wtdcode/GLM-5.3-Flash-AWQ-W4A16` (downloaded via hf-mirror, file sizes verified against the HF API — not a download issue)
- Serving: TP=8, `NCCL_P2P_DISABLE=1` (required, otherwise NCCL init hangs on this box), `NCCL_ALGO=Ring NCCL_PROTO=Simple`, `--disable-custom-all-reduce`
- Tested matrix: with/without MTP-3, with/without prefix caching, eager vs FULL_AND_PIECEWISE cudagraphs — **corruption persists in every combination**

### Symptom 1: prompt-token corruption (deterministic, temp=0)

The model misreads the first 1-2 content tokens of its prompt:

| User content | What the model reports reading |
|---|---|
| `你好` | `"5"` (literally reads the number five) |
| `hello` | "message is empty" |
| `1+1=` | `"1"` followed by garbled tokens |
| longer prompts (≥4 tokens) | survive semantically |

Key evidence that the server-side input is **correct**: `prompt_tokens` returned by vLLM (13/14/17 for the cases above) exactly match a local `apply_chat_template` rendering, and the token sequence contains the user content. The corruption happens between token IDs and model perception.

### Symptom 2: doubled tokens in generation

At temperature=0, generation contains duplicated tokens: "未达85%的目标**目标**", "伺**伺**服过热", "累计**累计**". Reproducible; survives MTP/prefix-cache/eager toggling.

`repetition_penalty=1.05` eliminates the doubling (1.1 breaks thinking). This is our current mitigation.

### Symptom 3: random token injection

Output contains tokens that never appeared in the prompt, e.g. `数据lobster数据源` ("lobster" injected mid-sentence).

### Cross-checks on the same machine / image / launch args

1. **`deepseek-ai/DeepSeek-V4-Flash-Vision-Exp` (official weights): ZERO corruption** on the identical probe set (short-prompt perception, doubling scan, agent-style long prompts). All 16 agent-readiness tests pass, numbers read precisely. → the fork's sm89 hybrid-arch path is not universally broken.
2. `canada-quant/GLM-5.3-Flash-W4A16-MTP` (GPTQ g128, experts-only): **different, worse failure** — generation collapses into multilingual gibberish loops in every config. Their README admits "≈37% deficit is SM120 kernel numerics" on non-Hopper parts.

→ On sm89, both third-party 4-bit quantizations of GLM-5.3-Flash are broken while the official-weights DeepSeek is clean.

### Hypothesis: expert under-calibration

The published `awq_run_info.json` shows:

```json
{"num_samples": 256, "seq_len": 2048, "calibrate_all_experts": false, "seconds": 3549}
```

With 288 routed experts × 8 active per token, 256 samples cannot cover all experts. Experts that received little/no calibration data would get unreliable scales — and the failure signature (specific token patterns hitting specific experts → corrupted logits for those patterns only) matches an expert-coverage hole rather than a global kernel problem.

### Ask

Would a recalibration with `calibrate_all_experts=true` + more/Chinese-heavier samples be feasible on your side? We're happy to run our repro suite (short-prompt perception probe, doubling scan, agent-style loop probe) on any candidate build on this L20 rig and report results.

Repro probe (fails on our box, passes on DeepSeek official):

```python
# vLLM OpenAI API, temperature=0
for p in ["你好", "hello", "1+1="]:
    r = client.chat.completions.create(model="glm-5.3-flash",
        messages=[{"role":"user","content":p}], max_tokens=300, temperature=0)
    # GLM-AWQ: model reports reading "5" / empty / "1"+garble
```

Thanks for the fork — it's the only way to run these frontier models on pre-Hopper parts, and the GLM line is tantalizingly close to usable on sm89.

---

## 中文摘要

我们的环境是 8×L20 (sm89)，v0.13.1 镜像（latest-sm89 与 pinned 版同 digest）。GLM-5.3-Flash-AWQ-W4A16 存在三类确定性损坏：① prompt 开头 1-2 个内容 token 被错误替换（"你好"→读到"5"，服务端 prompt_tokens 与本地模板逐一致，排除输入问题）；② 生成出现成对重复（"目标目标"，temp=0 可复现，与 MTP/前缀缓存/eager 无关）；③ 随机 token 注入（正文出现从未输入过的 "lobster"）。同机同镜像同参数下 DeepSeek-V4-Flash-Vision 官方权重完全干净，canada-quant 的 GPTQ 版则是另一种全灭。我们怀疑与 `calibrate_all_experts: false` + 256 样本不足以覆盖 288 专家有关。希望能出一版全专家校准的重建，我们可以在 L20 上跑完整的复现与验证套件配合测试。
